add streaming access log API

mattklein123 opened the pull request:
Signed-off-by: Matt Klein <mklein@lyft.com>

mattklein123 commented:
@ggreenway @htuch @twoism @rodaine @goaway

@htuch I'm a little confused on what I need to do to fix the build. Will work on that but the API can be reviewed.

mattklein123 commented:
@mrice32 @dnoe

mattklein123 commented:
> Should there be a way to specify which logs to receive? Basically something analogous to the configuration for file access logging. A format string is specified to tell what data is wanted, and then a file path as a destination. So a single envoy could have many separate access log files going at once.

My assumption here is that this will just be a new access log provider, so it will work with the existing filtering mechanism. You do raise the point that we do need a way to "name" the access log that is being streamed and then match that up with the configuration. Let me look at that. 

I think in terms of filtering the actual data, we should just sent everything since it's proto. Any reason not to? I guess perf, but seems like complexity that is not worth it to me.

mattklein123 commented:
We can configure a log name via https://github.com/envoyproxy/data-plane-api/blob/master/api/filter/accesslog.proto#L108 potentially, or add it as a first class field in AccessLog. I will think about this more and work on it Monday, but I wanted to get it out there for initial discussion.

rshriram commented:
In general, +1 for the idea. Would it make sense to support something like syslog protocol or something better? Basically something that will enable people to readily point envoys to their existing logging providers.

mattklein123 commented:
> Should we reuse the constructs from http/access_log ? The filters, et al.

Yes, that is the way it works. This is just a provider that swaps in for the existing file implementation. All filters remain.

> Would it make sense to support something like syslog protocol or something better?

We need gRPC at Lyft (and I'm guessing others will want this also). If someone wants to come along and implement a different provider that's of course fine.

douglas-reid commented on api/accesslog.proto:
```diff
@@ -169,3 +169,30 @@ message EnvoyAccessLog {
+message EnvoyAccessLogResponse {}
+
+// Service for streaming access logs from Envoy to an access log server.
+service EnvoyAccessLogService {
```
pardon the newb question: is there a reason to prepend "Envoy" to all the message/service names?  The package name already includes "envoy".

mattklein123 replied:
No reason. I was just copying what was already in the file, but I agree it's not useful and since no one is using the existing proto yet I will change all the names.

htuch replied:
+1

rshriram commented:
Now that you are here @douglas-reid, what do you think of slowly offloading our metrics and log reporting aspect of the mixer filter into this API ?

htuch commented on api/accesslog.proto:
```diff
@@ -169,3 +169,30 @@ message EnvoyAccessLog {
+  // API for "critical" access logs in which Envoy will buffer access logs for some period of time
+  // until it gets an ACK so it could then retry. This API is designed for high throughput with the
+  // expectation that it might be lossy.
+  rpc StreamAccessLogs(stream EnvoyAccessLogMessage)
```
How does the server get configured in the bootstrap for this?

mattklein123 replied:
This will be a new well known name for a log handler. I will update the docs, etc. to make it more clear.

htuch commented on api/accesslog.proto:
```diff
@@ -169,3 +169,30 @@ message EnvoyAccessLog {
+  // subsequent messages. This is a performance optimization.
+  Node node = 1;
+  // The individual access log message.
+  EnvoyAccessLog log = 2;
```
I don't think there is a huge benefit in making this a `repeated`, the byte saving on the wire is probably tiny compared to the size of even a single log message, but throwing it out there for consideration.

mattklein123 replied:
I don't have a strong opinion this. I don't see Envoy every sending more than one log message at a time but we could put it in to future proof it if we wanted to buffer a few logs before sending for example.

htuch commented on api/BUILD:
```diff
@@ -164,6 +164,16 @@ api_proto_library(
+    has_services = 1,
+    deps = [
+        ":base",
+        "@com_google_protobuf//:timestamp_proto",
```
I think if you move this to https://github.com/envoyproxy/data-plane-api/blob/master/bazel/api_build_system.bzl#L31 you will fix the build issue. The idea is we centralize the WKT dependencies.

mattklein123 commented:
All, updated. I moved the new API next to FileAccessLog which I hope will be more clear on how to configure it. I also update comments, etc. LMK if this is better.

mattklein123 commented on api/accesslog.proto:
```diff
@@ -9,7 +9,7 @@ import "google/protobuf/struct.proto";
 import "google/protobuf/wrappers.proto";
 
-message EnvoyAccessLog {
+message AccessLog {
```
Should we just move this entire proto into filter/accesslog.proto? Seems odd to have it sitting here.

mattklein123 replied:
Would probably rename to `AccessLogEntry` or something also.

mattklein123 replied:
I thought about it and I'm just going to move the proto into the other file. I don't see the point in having a separate file. Will update in a bit.

douglas-reid commented:
@rshriram I currently have a proposal out for introducing such APIs into Mixer, partially for this purpose. I don't know that we will ever want to totally eliminate the attributes-based reporting in the filter, but that is a longer discussion. We can take that to a more appropriate forum.

mattklein123 commented:
Updated again. @dnoe @mrice32 this may effect your import but as this is not used anywhere yet in public code I'm hoping this is OK. I think it's a lot more clear to the end user what is going on now.

htuch commented on api/filter/accesslog.proto:
```diff
@@ -95,16 +258,63 @@ message FileAccessLog {
+
+  // Prefix data that will only be sent in the first message on the stream. This is effectively
+  // structured metadata and is a performance optimization.
+  Prefix prefix = 1;
```
It might make sense to make this less stream oriented in terminology, and call this something like `Identifier`. I'm thinking of a scenario in which we have on the wire or in some stored proto a collection of log messages from different Envoys, in which we attached identification information. This could be done with what you have there right now, but it feels like a stretch in terms of naming.

mattklein123 replied:
Sure will change.

htuch commented on api/filter/accesslog.proto:
```diff
@@ -95,16 +258,63 @@ message FileAccessLog {
+  // The friendly name of the access log to be returned in StreamAccessLogsMessage.Prefix. This
+  // allows the access log server to differentiate between different access logs coming from the
+  // same Envoy.
+  string log_name = 1;
```
Shouldn't there be a cluster name in this config?

mattklein123 replied:
Oops yes. I would have realized this when implementing! Thanks.

mattklein123 commented:
@htuch updated
